{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# mRNA Reconstruction using RNA-Seq Next-Generation Sequencing Data\n",
    "The goal of this project is to reconstruct mRNA from RNA-Seq data. <br>\n",
    "In this specific situation, the RNA-Seq was extracted from healthy patients and patients diagnosed with Type 2 Diabetes. <br>\n",
    "Thus, this project examines the preproinsulin mRNA (pre-precursor of the insulin hormone/protein) of pancreatic beta cells.<br>\n",
    "\n",
    "### How to Run:\n",
    "Check all the imports in the next cell and install all packages you are missing.<br>\n",
    "Then simply run the cells.<br>\n",
    "Smaller necessary files are included in this GitHub repository while other files will be downloaded from this notebook.<br>\n",
    "Download speeds from the European Bioinformatics Institute vary and could take a while.<br>\n",
    "Estimated time to completion: About a week<br>\n",
    "The final products of most interest are probably in these folders: Counts, Reconstruct, and Summary.<br>\n",
    "\n",
    "#### System Requirements:\n",
    "This project requires about 40 GBs of storage (NGS data are large files).<br>\n",
    "RAM storage will likely not be a large issue for most computers.<br><br>\n",
    "\n",
    "\n",
    "\n",
    "For more general information about this project including biological interpretations, see my Medium article:<br>\n",
    "https://daovang.medium.com/simple-reconstruction-of-mrna-from-next-generation-sequencing-rna-seq-c4faaa5da90d?source=friends_link&sk=ee61abdaa22a0773f5030d08d47de577<br>\n",
    "\n",
    "All reference preproinsulin data was obtained from NCBI:<br>\n",
    "https://www.ncbi.nlm.nih.gov/gene?term=INS%5BGene%5D%20AND%20%22Homo%20sapiens%22%5BOrganism%5D&cmd=DetailsSearch<br>\n",
    "\n",
    "All NGS data is obtained from the European Bioinformatics Institute:<br>\n",
    "https://medium.com/r/?url=https%3A%2F%2Fwww.ebi.ac.uk%2Farrayexpress%2Fexperiments%2FE-MTAB-5061%2F<br>\n",
    "\n",
    "The original purpose of the data can be found in this publication in Cell:<br>\n",
    "https://www.cell.com/cell-metabolism/fulltext/S1550-4131(16)30436-3<br><br>\n",
    "\n",
    "Version 1.0<br>\n",
    "11/5/2020<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio # This is the biopython package\n",
    "from Bio import SeqIO, pairwise2\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from urllib.request import urlretrieve, urlopen\n",
    "\n",
    "import urllib, gzip, shutil, os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading Reference Sequences:\n",
    "# All reference sequences are obtained from NCBI\n",
    "\n",
    "## Insulin Variant 1\n",
    "var_1_ref_mRNA_cDNA = SeqIO.read(\"Insulin Reference mRNA and Protein Sequences\\\\NM_000207.3.fa\", \"fasta\")\n",
    "\n",
    "## Insulin Variant 2\n",
    "var_2_ref_mRNA_cDNA = SeqIO.read(\"Insulin Reference mRNA and Protein Sequences\\\\NM_001185097.2.fa\", \"fasta\")\n",
    "\n",
    "## Insulin Variant 3\n",
    "var_3_ref_mRNA_cDNA = SeqIO.read(\"Insulin Reference mRNA and Protein Sequences\\\\NM_001185098.2.fa\", \"fasta\")\n",
    "\n",
    "## Insulin Variant 4\n",
    "var_4_ref_mRNA_cDNA = SeqIO.read(\"Insulin Reference mRNA and Protein Sequences\\\\NM_001291897.2.fa\", \"fasta\")\n",
    "\n",
    "\n",
    "## Preproinsulin Reference Sequence:\n",
    "#### Note: All variants mRNA create the same protein/protein sequence (preproinsulin)\n",
    "ref_protein = SeqIO.read(\"Insulin Reference mRNA and Protein Sequences\\\\NP_000198.1.fa\", \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' \n",
    "Downloading RNA-Seq/transcriptomics data from https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-5061/samples/?full=true:\n",
    "\n",
    "See for experiment details: https://data.humancellatlas.org/explore/projects/ae71be1d-ddd8-4feb-9bed-24c3ddb6e1ad \n",
    "    This experiment (\"Single-cell RNA-seq analysis of human pancreas from healthy individuals and type 2 diabetes patients\")\n",
    "    only examines the RNA found in single pancreatic cells using RNA-Seq.\n",
    "        Although the experiment does collect RNA from all pancreatic cell types, I am only interested in beta cells because\n",
    "        these cells are the only insulin producing cells.\n",
    "\n",
    "Samples Collected from 10 Individuals:\n",
    "    6 Healthy Individuals\n",
    "    4 Individuals with Type 2 Diabetes (T2D)\n",
    "Download Size = 9.44 GBs (compressed files)\n",
    "270 FASTQ Files = 36.9 GBs (extracted/decompressed files)\n",
    "'''\n",
    "\n",
    "# Downloading the metadata/summary of all the samples collected in the experiment:\n",
    "if os.path.exists('RNA-Seq') != True: ## Creating a RNA-Seq folder\n",
    "    os.mkdir('RNA-Seq')\n",
    "    \n",
    "urlretrieve('https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-5061/E-MTAB-5061.sdrf.txt', 'RNA-Seq\\\\metadata.txt')    \n",
    "\n",
    "sample_reference = pd.read_csv('RNA-Seq\\\\metadata.txt', sep='\\t')\n",
    "\n",
    "## Splicing only specific rows from the metadata:\n",
    "### The first filter is to make sure the quality is \"OK\":\n",
    "sample_reference = sample_reference[sample_reference['Characteristics[single cell well quality]'] == 'OK']\n",
    "\n",
    "### The second filter is to select only for beta cells (insulin producing cells):\n",
    "sample_reference = sample_reference[sample_reference['Characteristics[cell type]'] == 'beta cell']\n",
    "\n",
    "\n",
    "\n",
    "# Downloading FASTQ files using the spliced metadata table:\n",
    "## Note: This may take long since download speeds with servers that hold biological information are typically slow\n",
    "\n",
    "counter = 0  # Setting up a counter for the tracking\n",
    "for index, row in sample_reference.iterrows():\n",
    "    print(row['Source Name'])\n",
    "    \n",
    "    ### The following below is a copy-and-paste from: https://stackoverflow.com/questions/22676/how-do-i-download-a-file-over-http-using-python\n",
    "    #### This is from the solution that adds in a \"progress bar\", which is important in this case because of slow download\n",
    "    #### speeds--ensuring that the connection to the server is still running.\n",
    "    url = row['Comment[FASTQ_URI]']\n",
    "\n",
    "    file_name = 'RNA-Seq\\\\' + row['Source Name'] + '.fastq.gz'\n",
    "    u = urlopen(url)\n",
    "    f = open(file_name, 'wb')\n",
    "    meta = u.info()\n",
    "    file_size = int(dict(meta.items())['Content-length'])\n",
    "    print (\"Downloading: %s Bytes: %s\" % (file_name, file_size))\n",
    "\n",
    "    file_size_dl = 0\n",
    "    block_sz = 8192\n",
    "    while True:\n",
    "        buffer = u.read(block_sz)\n",
    "        if not buffer:\n",
    "            break\n",
    "\n",
    "        file_size_dl += len(buffer)\n",
    "        f.write(buffer)\n",
    "        status = r\"%10d  [%3.2f%%]\" % (file_size_dl, file_size_dl * 100. / file_size)\n",
    "        status = status + chr(8)*(len(status)+1)\n",
    "        print(status)\n",
    "\n",
    "    f.close()\n",
    "    \n",
    "    ## This extracting the compressed file:\n",
    "    ### The following below is a copy-and-paste from: https://stackoverflow.com/questions/31028815/how-to-unzip-gz-file-using-python\n",
    "    with gzip.open('RNA-Seq\\\\' + row['Source Name'] + '.fastq.gz', 'rb') as f_in:\n",
    "        with open('RNA-Seq\\\\' + row['Source Name'] + '.fastq', 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)\n",
    "    \n",
    "    ## Removing the compressed .gz file:\n",
    "    os.remove('RNA-Seq\\\\' + row['Source Name'] + '.fastq.gz') \n",
    "    \n",
    "    counter += 1\n",
    "    print(str(counter / len(sample_reference) * 100) + \"% Samples Downloaded\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Finding the RNA sequences that match the mRNA of Variant 3 for reconstructing the mRNA preproinsulin sequence for each\n",
    "individual.\n",
    "\n",
    "Because the RNA-Seq reads are all in a length of 43 nucleotides, there are 3 possiblities where the reads can be\n",
    "matched in the preproinsulin mRNA reference sequence (644 nucleotides):\n",
    "\n",
    "    Also, indels (frameshifts) are not detected in this because the two final insulin chains (A and B) would most \n",
    "    likely NOT be functional as a hormone--the insulin receptor seems to bind to the entire insuline protein.\n",
    "        https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3793637/\n",
    "\n",
    "''' \n",
    "\n",
    "## Splitting the reference sequence of preproinsulin variant 3 by 20 nucleotides and creating a DataFrame for matches:\n",
    "variant_3 = str(var_3_ref_mRNA_cDNA.seq)\n",
    "var_3_length = len(var_3_ref_mRNA_cDNA.seq)  ## Number of nucleotides in variant 3\n",
    "\n",
    "## Grabbing all the FASTQ files downloaded:\n",
    "fastq_list = os.listdir('RNA-Seq\\\\')\n",
    "fastq_list.remove('metadata.txt')\n",
    "\n",
    "match_requirement = 35  ### Setting the matching requirement to be 35 nucleotides\n",
    "\n",
    "reconstruct_check_list = list(range(9, 42))\n",
    "reconstruct_check_list.extend(list(range(224, 634))) ### Checking for variant 1 exons, but skipping the first and last few\n",
    "\n",
    "\n",
    "if os.path.exists('Alignments') != True: ## Creating a RNA-Seq folder\n",
    "    os.mkdir('Alignments')\n",
    "\n",
    "for file in fastq_list:\n",
    "    fastq = SeqIO.parse('RNA-Seq\\\\' + file, \"fastq\")\n",
    "    print(file)\n",
    "    \n",
    "    reconstruct_df = pd.DataFrame(columns= list(range(0, 644))) ## Creating a DataFrame to store the matches\n",
    "    \n",
    "    index_counter = 0\n",
    "    reconstruct_counter = 0\n",
    "    for seq_record in fastq:\n",
    "        seq = str(seq_record.seq)\n",
    "        \n",
    "        matches = pairwise2.align.localms(variant_3, seq, 1, -1, -99, -99)  ### Matching & Aligning\n",
    "        \n",
    "        if matches[0].score >= match_requirement:  ### Aligning and saving to DataFrame\n",
    "            pair = matches[0]  ### Only taking the first alignment\n",
    "            \n",
    "            if \"-\" not in pair.seqA: ### \n",
    "                match_seq = list(pair.seqB)\n",
    "                start_index = pair.start\n",
    "                end_index = pair.end\n",
    "\n",
    "                match_index = start_index            \n",
    "                for z in range(start_index, end_index):\n",
    "                    align_seq = match_seq[match_index]\n",
    "                    \n",
    "                    if align_seq == '-':  ### Gaps are replaced with np.NaN\n",
    "                        align_seq = np.NaN\n",
    "                        \n",
    "                    reconstruct_df.loc[index_counter, match_index] = align_seq\n",
    "                    match_index += 1\n",
    "                \n",
    "                index_counter += 1\n",
    "                reconstruct_counter += 1\n",
    "                                \n",
    "        if reconstruct_counter == 1000:  ### Every 1000 matches, checking if each aligned nucleotide has at least 25 matches\n",
    "            finish_check = 0\n",
    "            print('check')\n",
    "            for u in reconstruct_check_list:\n",
    "                if len(reconstruct_df[u].dropna()) >= 25:  \n",
    "                    finish_check += 1\n",
    "                else:\n",
    "                    reconstruct_counter = 0\n",
    "                    break\n",
    "        \n",
    "            ### If all of the columns have at least 25 matches, checking through the file is stopped early:\n",
    "            if finish_check == len(reconstruct_check_list):  \n",
    "                print('exit')\n",
    "                break\n",
    "                    \n",
    "    align_filename = file.strip('.fastq')\n",
    "    align_filename = 'Alignments\\\\' + align_filename + '.csv'\n",
    "                \n",
    "    reconstruct_df.to_csv(align_filename) \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Now I want to:\n",
    "    1)\n",
    "        Count the number of nucleotides for each aligned matches\n",
    "    \n",
    "    2) \n",
    "        Determine the variant(s) produced by the individual\n",
    "    \n",
    "    3)\n",
    "        Reconstruct the possible mRNA preproinsulin for each individual\n",
    "        \n",
    "'''\n",
    "\n",
    "# Creating a reference of the 10 individuals:\n",
    "individual_ref = sample_reference[['Characteristics[individual]', 'Characteristics[disease]']].drop_duplicates()\n",
    "\n",
    "# Getting all the filenames saved in the previous cell:\n",
    "align_list = os.listdir('Alignments\\\\')\n",
    "\n",
    "if os.path.exists('Counts\\\\') != True: ## Creating a Counts folder\n",
    "    os.mkdir('Counts\\\\')\n",
    "    \n",
    "if os.path.exists('Reconstruct\\\\') != True: ## Creating a Reconstruct folder\n",
    "    os.mkdir('Reconstruct\\\\')\n",
    "\n",
    "for index, row in individual_ref.iterrows():\n",
    "    individual = row['Characteristics[individual]']\n",
    "    print(individual)\n",
    "    \n",
    "    concat_df = pd.DataFrame()\n",
    "    \n",
    "    for alignment_file in align_list:\n",
    "        if individual in alignment_file:\n",
    "            temp_df = pd.read_csv('Alignments\\\\' + alignment_file, index_col= 0, dtype= 'str')  \n",
    "                ## dtype= 'str' is needed because there are np.NaN and strings as values\n",
    "            \n",
    "            if len(concat_df) == 0:\n",
    "                concat_df = temp_df\n",
    "            \n",
    "            else:\n",
    "                concat_df = pd.concat([concat_df, temp_df], ignore_index= True)\n",
    "                \n",
    "    ## Creating a DataFrame to count the nucleotides for each position\n",
    "    nucleotide_df = pd.DataFrame(columns= list(range(0, 644)), index= ['A', 'T', 'G', 'C']) \n",
    "        \n",
    "    for column in concat_df.columns:\n",
    "        A_count = len(concat_df[concat_df[column] == 'A'])  ### The len() function works around needing to change np.NaN\n",
    "        T_count = len(concat_df[concat_df[column] == 'T'])\n",
    "        G_count = len(concat_df[concat_df[column] == 'G'])\n",
    "        C_count = len(concat_df[concat_df[column] == 'C'])\n",
    "        \n",
    "        nucleotide_df.loc['A', int(column)] = A_count  ### Conversion of \"column\" to int is needed because the column names  \n",
    "        nucleotide_df.loc['T', int(column)] = T_count  ### in nucleotide_df are int, while when temp_df was loaded, the \n",
    "        nucleotide_df.loc['G', int(column)] = G_count  ### column names were converted to strings\n",
    "        nucleotide_df.loc['C', int(column)] = C_count\n",
    "        \n",
    "    nucleotide_df.to_csv('Counts\\\\' + individual + '.csv')\n",
    "    \n",
    "    # Reconstructing the mRNA for all variants:\n",
    "    variant_df = pd.DataFrame(columns= list(range(0, 644)))\n",
    "    \n",
    "    variant_1_list = list(range(0, 42)) ## Variant 1, Exon 1\n",
    "    variant_1_list.extend(list(range(224, 644)))\n",
    "    \n",
    "    variant_2_list = list(range(0, 67)) ## Variant 2, Exon 1\n",
    "    variant_2_list.extend(list(range(224, 644)))\n",
    "    \n",
    "    variant_4_list = list(range(0, 101)) ## Variant 4, Exon 1\n",
    "    variant_4_list.extend(list(range(224, 644)))\n",
    "    \n",
    "    variant_3_list = list(range(0, 644)) ## Variant 3: Remember that this is the one used as a reference for alignments\n",
    "    \n",
    "    variant_list = [variant_1_list, variant_2_list, variant_3_list, variant_4_list]\n",
    "    \n",
    "    variant_reconstruct_df = pd.DataFrame(columns= list(range(0, 644)), \n",
    "                                          index= ['variant_1', 'variant_2', 'variant_3', 'variant_4'])\n",
    "    \n",
    "    variant_counter = 0\n",
    "    for variant in variant_list:\n",
    "        for nucleotide_position in variant:\n",
    "                ## Returning the index (A, T, G, or C) of the max count in that column\n",
    "                max_count = nucleotide_df[nucleotide_position].max()\n",
    "                \n",
    "                if max_count == 0:\n",
    "                    freq_nucleotide = ''\n",
    "                    variant_reconstruct_df.iloc[variant_counter, nucleotide_position] = freq_nucleotide\n",
    "                else:\n",
    "                    freq_nucleotide = nucleotide_df[nucleotide_df[nucleotide_position] == max_count].index\n",
    "                    variant_reconstruct_df.iloc[variant_counter, nucleotide_position] = freq_nucleotide[0]\n",
    "        \n",
    "        variant_counter += 1\n",
    "    \n",
    "    variant_reconstruct_df.to_csv('Reconstruct\\\\' + individual + '.csv')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding all the differences in mRNA sequences\n",
    "variant_3_list = list(str(var_3_ref_mRNA_cDNA.seq))\n",
    "counter = 0\n",
    "for x in variant_3_list:\n",
    "    compare_df.loc['reference', str(counter)] = x\n",
    "    counter += 1\n",
    "\n",
    "if os.path.exists('Summary\\\\') != True: ## Creating a Summary folder\n",
    "    os.mkdir('Summary\\\\')\n",
    "    \n",
    "compare_df.to_csv('Summary\\\\mRNA Comparison - Unfilled.csv')\n",
    "\n",
    "# Finding which nucleotide positions are different\n",
    "difference_df = pd.DataFrame()\n",
    "\n",
    "for nucleotide in list(range(0,644)):\n",
    "    temp_splice = compare_df[[str(nucleotide)]]\n",
    "    if len(set(compare_df[str(nucleotide)])) != 1:\n",
    "        if len(difference_df) == 0:\n",
    "            difference_df = temp_splice\n",
    "        else:\n",
    "            difference_df = pd.concat([difference_df, temp_splice], axis= 1)\n",
    "            \n",
    "difference_df.to_csv('Summary\\\\mRNA Comparison - Difference.csv')\n",
    "\n",
    "# Filling any empty nucleotides positions with the reference mRNA sequence\n",
    "compare_df_filled = compare_df.copy()\n",
    "for nucleotide in list(range(0,644)):\n",
    "    check_list = list(compare_df[str(nucleotide)])\n",
    "    while \"\" in check_list:\n",
    "        empty_position = list(compare_df[str(nucleotide)]).index(\"\")\n",
    "        compare_df_filled.iloc[empty_position, nucleotide] = compare_df_filled.loc[\"reference\", str(nucleotide)]\n",
    "        check_list = list(compare_df_filled[str(nucleotide)])\n",
    "        \n",
    "compare_df_filled.to_csv('Summary\\\\mRNA Comparison - Filled.csv')\n",
    "\n",
    "# Translating the mRNA sequences to protein sequences\n",
    "compare_protein_df = pd.DataFrame()\n",
    "\n",
    "variant_1_list = list(range(238, 571)) ## Variant 1 Protein Coding Region (same region for all other varients)\n",
    "variant_1_list = [str(x) for x in variant_1_list]\n",
    "\n",
    "for patient in list(compare_df_filled.index):\n",
    "    temp_string = \"\"\n",
    "\n",
    "    for nucleotide in list(compare_df_filled.loc[patient, variant_1_list]):\n",
    "        temp_string += nucleotide\n",
    "        \n",
    "    protein_seq = str(Seq(temp_string).translate())\n",
    "    \n",
    "    for column in list(range(len(protein_seq))):\n",
    "        compare_protein_df.loc[patient, column] = protein_seq[column]\n",
    "        \n",
    "compare_protein_df.to_csv('Summary\\\\Protein Comparison.csv')\n",
    "        \n",
    "# Comparing the protein sequences to find any differences\n",
    "difference_protein_df = pd.DataFrame()\n",
    "for aminoacid in list(range(110)):\n",
    "    temp_splice = compare_protein_df[[(aminoacid)]]\n",
    "    if len(set(compare_protein_df[(aminoacid)])) != 1:\n",
    "        if len(difference_protein_df) == 0:\n",
    "            difference_protein_df = temp_splice\n",
    "        else:\n",
    "            difference_protein_df = pd.concat([difference_protein_df, temp_splice], axis= 1)\n",
    "            \n",
    "difference_protein_df.to_csv('Summary\\\\Protein Comparison - Difference.csv')  ## There are no differences in protein sequences"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
